Rust/src/sampled_tree.rs

```diff
@@ -0,0 +1,267 @@
+extern crate num_traits;
+use num_traits::Float;
+
+use std::cell::{Ref, RefCell, RefMut};
+use std::iter::Sum;
+use std::rc::Rc;
+
+use crate::{PointStore, SamplerResult, StreamSampler};
+use crate::tree::{AddResult, NodeTraverser, Tree};
+
+/// Combination of a tree and a reservoir sampler.
+///
+/// A random cut tree, represented by the [`Tree`] struct, is a data structure
+/// for organizing vectors using bounding boxes and random cuts. In a random
+/// cut forest model we wish to limit the number of points contained in the
+/// tree. This is done using a reservoir sampler. A `SampledTree` coordinates
+/// point additions and deletions between a reservoir sampler, given by a
+/// [`StreamSampler`], and the random cut tree.
+///
+/// # Examples
+///
+/// ```
+/// use random_cut_forest::{Node, SampledTree};
+///
+/// // create a sampled tree that can hold 32 points witha a decay factor 0.01
+/// let mut tree: SampledTree<f32> = SampledTree::new(32, 0.01);
+///
+/// // if necessary, we can set the random seed used in point sampling as well
+/// // as the random cut process
+/// tree.seed(42);
+///
+/// // update the tree with some points. every point will be accepted into the
+/// // tree until the sample size is reached.
+/// tree.update(vec![0.0, 0.0], 0);
+/// tree.update(vec![1.0, -1.0], 1);
+/// tree.update(vec![2.0, 3.0], 2);
+/// assert_eq!(tree.num_observations(), 3);
+///
+/// // given a query point, we can traverse the tree by following random cuts
+/// let query = vec![0.5, 1.5];
+/// let traversal_nodes: Vec<&Node<f32>> = tree.traverse(&query).collect();
+/// ```
+pub struct SampledTree<T> {
+    point_store: Rc<RefCell<PointStore<T>>>,
```

Comment:
Good point about read-only vs. read and write access. 

Under the current design, both `SampledTree` and `Tree` perform mutable operations on the point store. (See `SampledTree::update()` for how this is done.) In a sense, the `SampledTree` is the primary owner of the point store and is offering up a cloned reference to the tree so it too can make changes.

An alternative is to make `Tree`s only communicate desired point store actions to `SampledTree`. This is probably what happens in the Java version, yes? Here, the `SampledTree` would perform the actual update. In a future version where the point store is actually shared across multiple trees, some forest-wide coordinator would perform the update.

I will certainly rework the logic when we get to the shared point store step of this work. (V1 to this V0.) The question is, how much of this approach should I start implementing now? The work would probably be to make trees only communicate intent rather than perform actual operations.

Reply:
In Java we have the "Update Coordinator" that sits above the individual models and manages the writes the shared point store. Our version of `SampledTree` accepts a point reference and returns an update result, and then the collection of all update results from individual trees are applied by the coordinator. The reason for this approach was to avoid the overhead of creating a point store that manages concurrent writes. I think this pattern could also simplify the ownership story in rust (but I think it's fine to address it in a future iteration).

Reply:
Another "that's future Chris's problem". 😃